### DISCUSSION TITLE
article: Why you shouldn't invoke setup.py directly

### DISCUSSION THREAD

**Discussion by ana** [avocado-framework/avocado#5051] [RFC]
This is an article that would be interested to keep in mind in future evolutions of `setup.py`:
https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html

This is a summary in the end https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html#summary

**Comment by richtja** [comment]
Just to be clear, that I understand it correctly. Right now, we are making a lot of effort to move from `Makefile` to `setup.py` but this article is about going further and stop using the `setup.py` also. So if we will decide to use approach from this article, shouldn't we stop the work on `Makefile` and `setup.py` and start thinking how to move away form `Makefile` directly with this new approach, am I right?

**Reply by beraldoleal** [reply]
We are doing a lot of effort to remove the lack of standard with regards Makefile vs setup.py. 

> This does not mean that setuptools itself is deprecated, or that using setup.py to configure your package builds is going to be removed.

setup.py is still used, in fact multiple commands that are the "suggested way" reads from setup.py in order to execute (ie. `pip install`). The article says that we should avoid calling directly. But it is still relevant.

So, IMO we should keep working on that, but taking in account those suggestions to make sure that we are using those suggestions.

**Reply by beraldoleal** [reply]
Another interesting reading about this setuptools / disutils mess can be found at PEP517.

**Reply by ana** [reply]
For me, it doesn't change anything of what we're doing now. But we must keep this article in mind for the long term.
In the short term (~9-12 months) I only see two things we should work on:
* Stop using distutils. This means re-writing 'clean' command, and considering that we not longer build the package locally and you can clean with git, this might not even be necessary. 
* A lot of of stuff from `setup.py/setup() ` can be moved to a declarative `setup.cfg`  file, see https://setuptools.pypa.io/en/latest/userguide/declarative_config.html

**Reply by beraldoleal** [reply]
I agree with you here @ana. Based on the current mess that is python dist vs setuptools I would give more time so we can see other early adopters and more community involvement. So yes, lets just keep doing what we are doing with the article in mind.

> Stop using distutils. This means re-writing 'clean' command, and considering that we not longer build the package locally and you can clean with git, this might not even be necessary.

I started already an effort on that front removing old distutils code and the only left it was with the 'clean' command because of lack of better choice since I avoiding rewriting-it at the time.

About the need, I would say that it is still useful. A simple `make develop` (or `pip install -e .` as suggested in the article) would still produce some temporary artifacts that some might want to get rid of. I would leave the removal of this command only after we are sure no temporary command is generated.

> A lot of of stuff from setup.py/setup() can be moved to a declarative setup.cfg file, see https://setuptools.pypa.io/en/latest/userguide/declarative_config.html

+1 for that.
